src/ripple/rpc/handlers/AccountTx.cpp

```diff
@@ -25,178 +25,633 @@
+            if (std::holds_alternative<LedgerSequence>(*ledgerSpecifier))
+            {
+                status = getLedger(
+                    ledgerView,
+                    std::get<LedgerSequence>(*ledgerSpecifier),
+                    context);
+            }
+            else if (std::holds_alternative<LedgerHash>(*ledgerSpecifier))
+            {
+                status = getLedger(
+                    ledgerView,
+                    std::get<LedgerHash>(*ledgerSpecifier),
+                    context);
+            }
+            else
+            {
+                assert(
+                    std::holds_alternative<LedgerShortcut>(*ledgerSpecifier));
+                status = getLedger(
+                    ledgerView,
+                    std::get<LedgerShortcut>(*ledgerSpecifier),
+                    context);
+            }
+            if (!ledgerView)
+            {
+                return status;
+            }
+
+            bool validated = RPC::isValidated(
+                context.ledgerMaster, *ledgerView, context.app);
 
-            for (auto& it: txns)
+            if (!validated || ledgerView->info().seq > uValidatedMax ||
+                ledgerView->info().seq < uValidatedMin)
             {
-                Json::Value& jvObj = jvTxns.append (Json::objectValue);
+                return rpcLGR_NOT_VALIDATED;
+            }
+            uLedgerMin = uLedgerMax = ledgerView->info().seq;
+        }
+    }
+    return LedgerRange{uLedgerMin, uLedgerMax};
+}
 
-                jvObj[jss::tx_blob] = std::get<0> (it);
-                jvObj[jss::meta] = std::get<1> (it);
+std::pair<AccountTxResult, RPC::Status>
+doAccountTxHelp(RPC::Context& context, AccountTxArgs& args)
+{
+    AccountTxResult result;
+    context.loadType = Resource::feeMediumBurdenRPC;
 
-                std::uint32_t uLedgerIndex = std::get<2> (it);
+    auto lgrRange = getLedgerRange(context, args.ledger);
+    if (std::holds_alternative<RPC::Status>(lgrRange))
+    {
+        // An error occurred getting the requested ledger range
+        return {result, std::get<RPC::Status>(lgrRange)};
+    }
+
+    result.ledgerRange = std::get<LedgerRange>(lgrRange);
+
+
+    if (args.binary)
+    {
+        result.transactions = context.netOps.getTxsAccountB(
+            args.account,
+            result.ledgerRange.min,
+            result.ledgerRange.max,
+            args.forward,
+            args.marker,
+            args.limit,
+            isUnlimited(context.role));
+    }
+    else
+    {
+        result.transactions = context.netOps.getTxsAccount(
+            args.account,
+            result.ledgerRange.min,
+            result.ledgerRange.max,
+            args.forward,
+            args.marker,
+            args.limit,
+            isUnlimited(context.role));
+    }
+
+    result.limit = args.limit;
+    if (args.marker)
+        result.marker = args.marker;
+
+    return {result, rpcSUCCESS};
+}
+
+void
+populateResponse(
```

Comment:
I tried to decouple the processing of the intermediate data (`AccountTxResult`), from the serialization of the response to the desired interface. I think these two tasks are logically separate from each other. What if we want to have a third interface type besides JSON and protobuf? The populate function makes it super easy to just add another interface type, where a developer only needs to define how to serialize to the new interface. For now, it seems not so bad to just write two separate yet very similar functions, but I think the `populateResponse` function ultimately makes the code more maintainable. It checks for null in the appropriate places, and the compiler forces the programmer to handle all of the pieces of the result data (for example, forgetting to do something with the marker will result in a compiler error).

However, I don't want to write code that is hard to understand, because that would make the code less maintainable.

Reply:
A brief comment from the peanut gallery.

I suggest caution about planning to support a third serialization format if it adds code complexity right now.  Future desired changes are likely to be things we don't anticipate.  If we really do need to add a third serialization format in the future, then knowing what that serialization form is will likely allow you to make a more appropriate change at that time.

If you can support additional serialization formats without adding complexity now, that would be a different story.  However it doesn't sound like that is the case.

Returning to lurk mode.

Reply:
I'm OK if you want to keep them; I do see some benefit. But I my vote is I'd rather make the code easier to read.

Reply:
I've been swayed. Making the code easy to read is always important, and I think the benefits of having one function and passing lambdas are not enough to compromise the readability here. Will change